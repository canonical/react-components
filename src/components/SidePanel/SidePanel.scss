@import "vanilla-framework";

.side-panel.is-overlay {
  position: absolute;
  right: 0;
  z-index: 103 !important;
}

.side-panel {
  background-color: $colors--theme--background-default;
  height: 100dvh;
  padding: 0 $sph--x-large;
  padding-bottom: $spv--medium;
  padding-top: $spv--small;

  .error,
  .loading {
    padding-top: $spv--medium;
  }

  .p-panel__header {
    padding: 0;
  }

  .panel-footer {
    .rule {
      margin-bottom: $spv--medium;
    }
  }

  .sticky-wrapper {
    background: $colors--theme--background-default;
    position: sticky;
    top: 0;
    z-index: 1;
  }

  .sticky-wrapper--top {
    top: 0;
  }

  .sticky-wrapper--bottom {
    bottom: 0;
  }
}
