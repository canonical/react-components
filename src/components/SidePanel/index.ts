export { default, type Props as SidePanelProps } from "./SidePanel";
