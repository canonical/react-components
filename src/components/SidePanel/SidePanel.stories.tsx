import React, { useEffect, useId, useRef, useState } from "react";
import type { Meta, StoryObj } from "@storybook/react";

import SidePanel from "./SidePanel";
import Button from "../Button";
import Application from "../ApplicationLayout/Application";
import AppMain from "../ApplicationLayout/AppMain";
import { ActionButton } from "../../index";
import Icon from "../Icon";

const meta: Meta<typeof SidePanel> = {
  component: SidePanel,
  tags: ["autodocs"],
  decorators: [
    (Story) => {
      const ref = useRef<HTMLDivElement>(null);

      useEffect(() => {
        const el = ref.current;
        if (el) {
          const wrapper = el?.parentElement?.parentElement?.parentElement;
          const parent = wrapper?.parentElement?.parentElement?.parentElement;
          console.log(wrapper);
          if (wrapper) {
            wrapper.style.setProperty("border", "0", "important");
          }
          if (parent) {
            parent.style.margin = "0";
            parent.style.padding = "0";
          }
        }
      }, []);
      document.body.style.padding = "0";

      return (
        <div ref={ref}>
          <Story />
        </div>
      );
    },
  ],
};

export default meta;

type Story = StoryObj<typeof meta>;

const StoryExample = (args: Story["args"]) => {
  const [isOpen, setIsOpen] = useState(false);

  // Multiple stories get rendered on the docs page, ensure the ids are not colliding
  const id = useId();
  const parentId = args.parentId + id;

  const openPanel = () => {
    setIsOpen(true);
  };

  const closePanel = () => {
    setIsOpen(false);
  };

  // Delay opening the panel, so the l-application container is ready to receive it
  useEffect(() => void setTimeout(openPanel, 100), []);

  return (
    <Application id={parentId}>
      <AppMain>
        <Button onClick={openPanel} style={{ margin: "1rem" }}>
          Open side panel
        </Button>
        <Button onClick={closePanel} style={{ margin: "1rem" }}>
          Close side panel
        </Button>
      </AppMain>
      {isOpen && (
        <SidePanel
          overlay={args.overlay}
          loading={args.loading}
          hasError={args.hasError}
          parentId={parentId}
          pinned={args.pinned}
          wide={args.wide}
          className="u-no-padding--top u-no-padding--bottom"
        >
          <SidePanel.Sticky>
            <SidePanel.Header>
              <SidePanel.HeaderTitle>Edit panel</SidePanel.HeaderTitle>
              <SidePanel.HeaderControls>
                <Button
                  appearance="base"
                  className="u-no-margin--bottom"
                  hasIcon
                  onClick={closePanel}
                  aria-label="Close"
                >
                  <Icon name="close" />
                </Button>
              </SidePanel.HeaderControls>
            </SidePanel.Header>
          </SidePanel.Sticky>
          <SidePanel.Content className="u-no-padding">
            <p>Here be dragons!</p>
            <p>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras
              varius mi eu pretium vulputate. Nunc commodo sit amet nibh quis
              rhoncus. Aliquam rhoncus porttitor semper. Aenean faucibus
              consectetur neque in sodales. Sed cursus mauris id ex sollicitudin
              sodales. Quisque molestie rutrum odio, ornare pharetra ligula.
              Pellentesque ornare tristique feugiat. In a augue neque. Aenean
              eget arcu quis lacus tempus posuere in sit amet dui. Suspendisse
              faucibus sapien nisl, nec laoreet sem convallis nec.
            </p>
            <p>
              Vestibulum sed placerat lorem. Nam luctus ex id nisi luctus, id
              vestibulum sem bibendum. Vivamus turpis sem, pellentesque
              fermentum malesuada eu, faucibus porta libero. Duis eget venenatis
              odio. Etiam sed volutpat magna, non tempus erat. Nunc id tortor ac
              quam consectetur dapibus ac ut tellus. Pellentesque ut tellus
              venenatis elit vehicula condimentum eget quis lorem. Ut non
              consectetur est, a fringilla ipsum. Nunc vitae ligula ipsum. Etiam
              suscipit, libero ut lacinia viverra, nunc urna consequat ex, vel
              eleifend eros mauris vitae ipsum. Pellentesque sed dictum augue.
              Ut sit amet ullamcorper mauris. Nunc congue orci mollis purus
              sodales facilisis ac ut arcu. Maecenas feugiat sapien ac massa
              mollis sodales. Donec vitae turpis eu nisi laoreet pulvinar quis
              at nisl. Integer volutpat, metus eget elementum dictum, lacus
              sapien viverra felis, consequat fermentum nisl mi ac dui.
            </p>
            <p>
              Nullam nulla turpis, dignissim vel dapibus ut, volutpat ac dui.
              Donec vel elementum lacus. Mauris maximus nec felis at faucibus.
              Nunc faucibus gravida velit, id blandit lectus tincidunt ac.
              Vestibulum orci diam, elementum in congue eu, placerat id risus.
              Sed tempor tempus tellus, vitae iaculis turpis fringilla nec.
              Phasellus imperdiet facilisis velit, sit amet lobortis odio
              dignissim ut.
            </p>
            <p>
              Nam placerat urna vitae ligula hendrerit, ac tincidunt lorem
              maximus. Mauris eu odio nisi. Nulla facilisi. Sed egestas elit sed
              velit rutrum, sit amet bibendum metus hendrerit. Pellentesque
              luctus placerat tellus, eu bibendum justo. Cras eget leo ac ex
              volutpat gravida. Duis vitae mollis ante. Duis a congue nunc.
              Aenean aliquet, sapien quis tincidunt tincidunt, odio eros
              consectetur lacus, vel finibus mauris tortor id velit. Donec
              tincidunt vitae purus eu interdum. Pellentesque scelerisque dui
              viverra ex ullamcorper volutpat. Vestibulum lacinia vitae arcu
              volutpat porta. Etiam et cursus nulla, id aliquet felis. Nam
              ultricies, urna id mattis pretium, velit erat viverra elit, eu
              maximus diam eros id nisi.
            </p>
            <p>
              Nullam eget nisl lectus. Pellentesque eu mauris ut tortor
              malesuada sagittis. Cras dictum cursus est non ultricies. Duis
              mollis non neque at commodo. Nunc feugiat justo et consequat
              aliquam. Ut consectetur libero eu erat feugiat finibus. Duis
              varius convallis quam eu sagittis. Maecenas ac est arcu.
              Suspendisse at enim eget nibh ultricies dictum. Etiam aliquet
              tellus vel felis malesuada laoreet.
            </p>
          </SidePanel.Content>
          <SidePanel.Sticky position="bottom">
            <SidePanel.Footer className="u-align--right">
              <Button appearance="base" onClick={closePanel}>
                Cancel
              </Button>
              <ActionButton appearance="positive">Save changes</ActionButton>
            </SidePanel.Footer>
          </SidePanel.Sticky>
        </SidePanel>
      )}
    </Application>
  );
};

export const Default: Story = {
  args: {
    className: "",
    hasError: false,
    parentId: "l-application-default",
    pinned: false,
    loading: false,
    overlay: true,
    wide: false,
  },
  render: StoryExample,
};

export const Pinned: Story = {
  args: {
    className: "",
    hasError: false,
    parentId: "l-application-pinned",
    pinned: true,
    loading: false,
    overlay: false,
    wide: false,
  },
  render: StoryExample,
  parameters: {
    docs: {
      disable: true,
    },
  },
};

export const Loading: Story = {
  args: {
    className: "",
    hasError: false,
    parentId: "l-application-loading",
    pinned: false,
    loading: true,
    overlay: true,
    wide: false,
  },
  render: StoryExample,
};

export const Error: Story = {
  args: {
    className: "",
    hasError: true,
    parentId: "l-application-error",
    pinned: false,
    loading: false,
    overlay: true,
    wide: false,
  },
  render: StoryExample,
};

export const Wide: Story = {
  args: {
    className: "",
    hasError: false,
    parentId: "l-application-wide",
    pinned: false,
    loading: false,
    overlay: true,
    wide: true,
  },
  render: StoryExample,
};
