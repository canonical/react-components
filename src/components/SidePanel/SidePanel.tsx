import React from "react";
import classnames from "classnames";
import { FC, PropsWithChildren, ReactNode } from "react";
import { createPortal } from "react-dom";
import Content, { type ContentProps } from "./common/Content";
import Footer, { type FooterProps } from "./common/Footer";
import Header, { type HeaderProps } from "./common/Header";
import HeaderControls, {
  type HeaderControlsProps,
} from "./common/HeaderControls";
import HeaderTitle, { type HeaderTitleProps } from "./common/HeaderTitle";
import Sticky, { type StickyProps } from "./common/Sticky";
import AppAside from "../ApplicationLayout/AppAside";
import Spinner from "../Spinner";
import "./SidePanel.scss";

/**
 * The props for the SidePanelComponent component.
 */
export type Props = {
  /**
   * The content of the side panel.
   */
  children: ReactNode;

  /**
   * Additional CSS classes to apply.
   */
  className?: string;

  /**
   * Whether the side panel has an error. This will show an error message in the panel instead of rendering the children.
   */
  hasError?: boolean;

  /**
   * Whether the side panel is currently loading. This will show a spinner in the panel instead of rendering the children.
   */
  loading?: boolean;

  /**
   * Whether the side panel is an overlay with background or should allow seeing the content below it.
   */
  overlay?: boolean;

  /**
   * The ID of the parent element to which the side panel will be appended
   */
  parentId?: string;

  /**
   * Whether the side panel should be pinned. When pinned, it will remain visible alongside the main content.
   */
  pinned?: boolean;

  /**
   * Whether the side panel should be wide.
   */
  wide?: boolean;
};

/**
This is a [React](https://reactjs.org/) component for the Vanilla application layout [Side Panel](https://vanillaframework.io/docs/layouts/application#panels).

The sidepanel component should be used to show additional information relating to the main content on the page. It is using a composable structure with building blocks:

* **SidePanel.Header:** Defines the header of the side panel, which can include a title and controls.

* **SidePanel.HeaderTitle:** To show the title of the side panel as heading.

* **SidePanel.HeaderControls:** To show controls in the header, such as buttons or icons for actions like closing the panel.

* **SidePanel.Sticky:** Can be wrapped around the header or footer to make them sticky when scrolling.

* **SidePanel.Content:** To show the main content of the side panel.

* **SidePanel.Footer:** To show additional information or actions at the bottom of the side panel.

## Example

Usage of the composable structure:

```jsx
<SidePanel className="u-no-padding--top u-no-padding--bottom">
  <SidePanel.Sticky>
    <SidePanel.Header>
      <SidePanel.HeaderTitle>Edit panel</SidePanel.HeaderTitle>
    </SidePanel.Header>
  </SidePanel.Sticky>
  <SidePanel.Content className="u-no-padding">
    <p>Here be dragons!</p>
  </SidePanel.Content>
  <SidePanel.Sticky position="bottom">
    <SidePanel.Footer className="u-align--right">
      <Button appearance="base">Cancel</Button>
      <ActionButton appearance="positive">Save changes</ActionButton>
    </SidePanel.Footer>
  </SidePanel.Sticky>
</SidePanel>
```
*/
const SidePanelComponent = ({
  children,
  className,
  hasError = false,
  loading = false,
  overlay,
  pinned,
  wide,
  parentId = "l-application",
}: Props): React.JSX.Element => {
  const container = document.getElementById(parentId) || document.body;

  return (
    <>
      {createPortal(
        <AppAside
          className={classnames("side-panel", className, {
            "is-overlay": overlay,
          })}
          aria-label="Side panel"
          pinned={pinned}
          narrow={!wide}
          wide={wide}
        >
          {loading ? (
            <div className="loading">
              <Spinner />
            </div>
          ) : hasError ? (
            <div className="error">Loading failed</div>
          ) : (
            children
          )}
        </AppAside>,
        container,
      )}
    </>
  );
};

type SidePanelComponents = FC<Props> & {
  Header: FC<PropsWithChildren & HeaderProps>;
  HeaderTitle: FC<PropsWithChildren & HeaderTitleProps>;
  HeaderControls: FC<PropsWithChildren & HeaderControlsProps>;
  Sticky: FC<PropsWithChildren & StickyProps>;
  Content: FC<PropsWithChildren & ContentProps>;
  Footer: FC<PropsWithChildren & FooterProps>;
};

const SidePanel = SidePanelComponent as SidePanelComponents;
SidePanel.Header = Header;
SidePanel.HeaderTitle = HeaderTitle;
SidePanel.HeaderControls = HeaderControls;
SidePanel.Sticky = Sticky;
SidePanel.Content = Content;
SidePanel.Footer = Footer;

export default SidePanel;
