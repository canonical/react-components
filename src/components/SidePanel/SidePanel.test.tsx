import React from "react";
import { render, screen } from "@testing-library/react";

import SidePanel from "./SidePanel";

it("renders", () => {
  render(<SidePanel>SidePanel</SidePanel>);
  expect(screen.getByText("SidePanel")).toBeInTheDocument();
});

it("applies className", () => {
  render(<SidePanel className="test-class">SidePanel</SidePanel>);
  expect(screen.getByText("SidePanel")).toHaveClass("test-class");
});

it("displays as loading", () => {
  render(<SidePanel loading>SidePanel</SidePanel>);
  expect(screen.getByText("Loading")).toBeInTheDocument();
  expect(screen.queryByText("SidePanel")).not.toBeInTheDocument();
});

it("displays with error", () => {
  render(<SidePanel hasError>SidePanel</SidePanel>);
  expect(screen.getByText("Loading failed")).toBeInTheDocument();
  expect(screen.queryByText("SidePanel")).not.toBeInTheDocument();
});

it("displays as pinned", () => {
  render(<SidePanel pinned>SidePanel</SidePanel>);
  expect(screen.getByText("SidePanel")).toHaveClass("is-pinned");
});

it("displays as wide", () => {
  render(<SidePanel wide>SidePanel</SidePanel>);
  expect(screen.getByText("SidePanel")).toHaveClass("is-wide");
});

it("assembles complex structure", () => {
  render(
    <SidePanel>
      <SidePanel.Sticky>
        <SidePanel.Header>
          <SidePanel.HeaderTitle>Edit panel</SidePanel.HeaderTitle>
        </SidePanel.Header>
      </SidePanel.Sticky>
      <SidePanel.Content className="u-no-padding">
        <p>Here be dragons!</p>
      </SidePanel.Content>
      <SidePanel.Sticky position="bottom">
        <SidePanel.Footer className="u-align--right">
          Footer content
        </SidePanel.Footer>
      </SidePanel.Sticky>
    </SidePanel>,
  );
  expect(
    screen.getByText("Here be dragons!").parentElement.parentElement,
  ).toMatchSnapshot();
});
