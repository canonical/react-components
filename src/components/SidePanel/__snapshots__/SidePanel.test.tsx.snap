// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assembles complex structure 1`] = `
<aside
  aria-label="Side panel"
  class="l-aside side-panel"
>
  <div
    class="sticky-wrapper sticky-wrapper--top"
  >
    <div
      class="p-panel__header"
    >
      <h2
        class="p-panel__title"
      >
        Edit panel
      </h2>
    </div>
  </div>
  <div
    class="p-panel__content u-no-padding"
  >
    <p>
      Here be dragons!
    </p>
  </div>
  <div
    class="sticky-wrapper sticky-wrapper--bottom"
  >
    <div
      class="panel-footer u-align--right"
      id="panel-footer"
    >
      <hr
        class="rule"
      />
      Footer content
    </div>
  </div>
</aside>
`;
